pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +351,446 @@ pub mod pallet {
+			ensure!(
+				ForeignRedemptionInfo::<T>::contains_key(&who, investment_id),
+				Error::<T>::InfoNotFound
+			);
+
+			T::Investment::redemption(who, investment_id)
+		}
+
+		fn accepted_payment_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payment_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
+					.unwrap_or(false)
+			}
+		}
+
+		fn accepted_payout_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payout_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+					.unwrap_or(false)
+			}
+		}
+	}
+
+	pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+		type Error = DispatchError;
+		type Id = T::SwapId;
+		type Status = SwapOf<T>;
+
+		fn notify_status_change(
+			swap_id: T::SwapId,
+			last_swap: SwapOf<T>,
+		) -> Result<(), DispatchError> {
+			let (who, investment_id) =
+				SwapIdToForeignId::<T>::get(swap_id).ok_or(Error::<T>::SwapOrderNotFound)?;
+
+			let pending_amount = match T::TokenSwaps::get_order_details(swap_id) {
+				Some(swap) => swap.amount_in,
+				None => {
+					SwapIdToForeignId::<T>::remove(swap_id);
+					T::Balance::zero()
+				}
+			};
+
+			ForeignInvestmentInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				if let Some(info) = maybe_info {
+					if info.increase_swap_id == Some(swap_id) {
+						T::Investment::update_investment(
+							&who,
+							investment_id,
+							T::Investment::investment(&who, investment_id)?
+								.ensure_add(last_swap.amount_in)?,
+						)?;
+
+						if pending_amount.is_zero() {
+							info.increase_swap_id = None;
+						}
+					} else if info.decrease_swap_id == Some(swap_id) {
+						info.decrease_swapped_amount
+							.ensure_add_assign(last_swap.amount_in)?;
+
+						if pending_amount.is_zero() {
+							// NOTE: How make this works with market ratios?
+							let remaining_foreign_amount = T::CurrencyConverter::stable_to_stable(
+								info.base.foreign_currency,
+								Pallet::<T>::pool_currency_of(investment_id)?,
+								T::Investment::investment(&who, investment_id)?,
+							)?;
+
+							T::DecreasedForeignInvestOrderHook::notify_status_change(
+								(who.clone(), investment_id),
+								ExecutedForeignDecreaseInvest {
+									amount_decreased: info.decrease_swapped_amount,
+									foreign_currency: info.base.foreign_currency,
+									amount_remaining: remaining_foreign_amount,
+								},
+							)?;
+
+							info.decrease_swapped_amount = T::Balance::default();
+							info.decrease_swap_id = None;
```

Comment:
IIUC, we reset all `swap` data here except for the `foreign_currency` such that any `(investor_address, investment_id)` pair right now is limited to their initial `foreign_currency` which is never reset, even if all swaps, investments and redemptions are cleared. 

There should be some condition which clears `ForeignInvestmentInfo`. IMO, this should be the case if there is neither a swap, nor a pending investment.

Reply:
Good point! I think you're right. I was clearing line `759` under the investment collected, but the user could decrease everything without collecting. In this case I should clean where you say.

Thanks for your insightful comments!

Reply:
Fixed. I think this is the correct check: [eb43c1f](https://github.com/centrifuge/centrifuge-chain/pull/1698/commits/eb43c1fa36394997bc5854594c5718c069152860), ensuring that we only remove if there is no more pending investment collects.